### ISSUE TITLE
How exactly is MapMessage supposed to be handled?

### ISSUE THREAD

**Issue by johnament** [vy/log4j2-logstash-layout#61]
Hi

I saw your bug report (and fix) for Log4J2's MapMessage handling.  However, I've tried with a few different logging approaches and none seem to give me the output I was hoping for (where the MapMessage is output as a JSON Object).

Is there a little bit better instruction somewhere of how a MapMessage can be handled, or is it a limitation that the MapMessage class can't be serialized to an object?

Ultimately what I'm trying to do is log all headers into separate fields in my JSON log for an incoming request (to make each field searchable as well as run aggregates on different header values).

**Comment by johnament** [comment]
Here's some resources.  My json log format:

```json
{
  "mdc": "${json:mdc}",
  "ndc": "${json:ndc}",
  "exception": {
    "exception_class": "${json:exception:className}",
    "exception_message": "${json:exception:message}",
    "stacktrace": "${json:exception:stackTrace:text}"
  },
  "line_number": "${json:source:lineNumber}",
  "class": "${json:source:className}",
  "logVersion": 1,
  "source_host": "${hostName}",
  "message": "${json:message}",
  "thread_name": "${json:thread:name}",
  "@appTimestamp": "${json:timestamp}",
  "level": "${json:level}",
  "levelNum": "${json:level:severity:code}",
  "file": "${json:source:fileName}",
  "method": "${json:source:methodName}",
  "logger_name": "${json:logger:name}",
  "headers": "${json:map:headers}"
}
```

as well as my log4j2 xml config
```xml
        <Console name="CONSOLE" target="SYSTEM_OUT">
            <LogstashLayout dateTimeFormatPattern="yyyy-MM-dd'T'HH:mm:ss.SSSZZZ"
                            eventTemplateUri="classpath:/logging/template.json"
                            prettyPrintEnabled="true"
                            emptyPropertyExclusionEnabled="true"
                            stackTraceEnabled="true"/>
        </Console>
```

**Comment by vy** [comment]
@johnament, you need to use `${json:message:json}` directive:

```java

LogstashLayout layout = LogstashLayout
        .newBuilder()
        .setConfiguration(new DefaultConfiguration())
        .setEventTemplate("{'message':'${json:message:json}'}".replaceAll("'", "\""))
        .build();

Message message1 = new MapMessage()
        .with("key1", "val1")
        .with("key2", "val2")
        .with("key3", Collections.singletonMap("foo", "bar"));
LogEvent logEvent1 = Log4jLogEvent
        .newBuilder()
        .setMessage(message1)
        .build();

String serializedLogEvent1 = layout.toSerializable(logEvent1);
Assertions
        .assertThat(serializedLogEvent1)
        .startsWith("{'message':{'key1':'val1','key2':'val2','key3':{'foo':'bar'}}}"
                .replaceAll("'", "\""));

Message message2 = new SimpleMessage("issue 61");
LogEvent logEvent2 = Log4jLogEvent
        .newBuilder()
        .setMessage(message2)
        .build();

String serializedLogEvent2 = layout.toSerializable(logEvent2);
Assertions
        .assertThat(serializedLogEvent2)
        .startsWith("{'message':'issue 61'}"
                .replaceAll("'", "\""));
```

Does this help?

**Comment by johnament** [comment]
@vy I had seen that, but then the behavior I was looking for ultimately wasn't coming through correctly.  While `${json:message:json}` does in fact work, it only works for the message field in the JSON.  TBH I would have assumed that the default rendering of a message if type is `ObjectMessage` or `MapMessage` would be as a JSON object instead of me having to explicitly list it.  Reason being is now a developer has to be careful to always use `MapMessage` or else `${json:message:json}` seems to render a nest object always (may not be a bad thing, still reviewing with our logging admins).

Ultimately though, what I was really trying to get at was in this section of the template:

```
"headers": "${json:map:headers}"
```

In my subclass, I'm actually using a `Map<String,Object>` which does in fact serialize properly when I use `${json:message:json}` but what I was hoping to do in my template is something where I can refer to all known map entry names (or if object message can be supported object fields) and have them serialize properly as JsonNodes from the Jackson perspective, resulting in a template looking like this (based on what you can support):

```
"httpHeaders": "${json:map:headers:json}",
"httpMethod":"${json:map:httpMethod:json}",
"message":"${json:map:message:json}",
"duration":"${json:map:duration:json}"
```

So not sure if it's worth putting in a ticket for an enhancement to treat these as json nodes or string elements when being processed, especially since I already see you're bringing changes into the ASF.

**Comment by vy** [comment]
Now I see your point! That indeed gave me a good insight on how broken the way we treat `MapMessage`s. I will draft a quick-and-dirty fix for you in `LogstashLayout` and come up with something better for `JsonTemplateLayout`. Give me a couple of days. Thanks so much again for the feedback.

**Comment by vy** [comment]
@johnament, I've pushed a83db2b which preserves the type of values in `MapMessage`s accessed via `map` directive. Consider the following test:

```java
// Create the log event.
MapMessage<?, ?> message = new StringMapMessage()
        .with("key1", "val1")
        .with("key2", Arrays.asList(1, 2));
LogEvent logEvent = Log4jLogEvent
        .newBuilder()
        .setLoggerName(LogstashLayoutTest.class.getSimpleName())
        .setLevel(Level.INFO)
        .setMessage(message)
        .build();

// Create the event template node with map values.
ObjectNode eventTemplateRootNode = JSON_NODE_FACTORY.objectNode();
eventTemplateRootNode.put("mapValue1", "${json:map:key1}");
eventTemplateRootNode.put("mapValue2", "${json:map:key2}");
eventTemplateRootNode.put("mapValue3", "${json:map:noExist}");
String eventTemplate = eventTemplateRootNode.toString();

// Create the layout.
Configuration configuration = ConfigurationBuilderFactory.newConfigurationBuilder().build();
LogstashLayout layout = LogstashLayout
        .newBuilder()
        .setConfiguration(configuration)
        .setEventTemplate(eventTemplate)
        .setEmptyPropertyExclusionEnabled(true)
        .build();

// Check serialized event.
String serializedLogEvent = layout.toSerializable(logEvent);
JsonNode rootNode = OBJECT_MAPPER.readTree(serializedLogEvent);
assertThat(point(rootNode, "mapValue1").asText()).isEqualTo("val1");
assertThat(point(rootNode, "mapValue2").isArray()).isTrue();
assertThat(point(rootNode, "mapValue2").get(0).asInt()).isEqualTo(1);
assertThat(point(rootNode, "mapValue2").get(1).asInt()).isEqualTo(2);
assertThat(point(rootNode, "mapValue3")).isInstanceOf(MissingNode.class);
```

Would you mind confirming if this addresses your issue, please?

**Comment by johnament** [comment]
Hi, yes that in fact does resolve the ask.  Will this also be in the log4j2 version of the library?

**Comment by vy** [comment]
I have just released v1.0.3 to Maven Central -- it should become visible in half an hour or so. Yes, I have indeed [incorporated the fix](/apache/logging-log4j2/commit/18750c1647421a83b6256b5df28447286c772e8b) into `JsonTemplateLayout` too. Thanks so much again for taking your time!
